lib/rubocop/cop/style/hash_transform_methods.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for uses of `_.each_with_object({}) {...}`,
+      # `_.map {...}.to_h`, and `Hash[_.map {...}]` that are actually just
+      # transforming either the keys or the values of a hash, and tries to use
+      # a simpler & faster call to `transform_keys` or `transform_values`
+      # instead.
+      #
+      # This can produce false positives if we are transforming an enumerable
+      # of key-value-like pairs that isn't actually a hash, e.g.:
+      # `[[k1, v1], [k2, v2], ...]`
+      #
+      # This cop should only be enabled on Ruby version 2.5 or newer
+      # (`transform_values` was added in 2.4, and `transform_keys` in 2.5.)
+      #
+      # @example
+      #   # bad
+      #   {a: 1, b: 2}.each_with_object({}) { |(k, v), h| h[k] = v * v }
+      #   {a: 1, b: 2}.map { |k, v| [k.to_s, v] }
+      #
+      #   # good
+      #   {a: 1, b: 2}.transform_values { |v| v * v }
+      #   {a: 1, b: 2}.transform_keys { |k| k.to_s }
+      class HashTransformMethods < Cop
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        EACH_WITH_OBJECT_PATTERN = NodePattern.new(<<~PATTERN)
+          (block
+            ({send csend} _ :each_with_object (hash))
+            (args
+              (mlhs
+                (arg $_)
+                (arg $_))
+              (arg _memo))
+            ({send csend} (lvar _memo) :[]= $_ $_))
+        PATTERN
+
+        HASH_BRACKETS_MAP_PATTERN = NodePattern.new(<<~PATTERN)
+          (send
+            (const _ :Hash)
+            :[]
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_)))
+        PATTERN
+
+        MAP_TO_H_PATTERN = NodePattern.new(<<~PATTERN)
+          ({send csend}
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_))
+            :to_h)
```

Comment:
Since Ruby 2.6 [`to_h` can take a block](https://ruby-doc.org/core-2.6/Enumerable.html#method-i-to_h), but this pattern would still match and produce a false positive:

```ruby
hash.map { |k, v| [k.to_s, v] }.to_h { |k, v| [v, k] }
```

I solved this by wrapping `[... !block_literal?]` around the whole pattern ([logical AND](https://rubocop.readthedocs.io/en/latest/node_pattern/#for-and)).

Reply:
Interesting, but I'm not totally convinced that should count as a false positive. Sticking with your example, feels like it would be cleaner as `hash.transform_keys(&:to_s).invert`, as this cop would suggest. Even if it wasn't a hash, I think I'd still prefer `not_a_hash.to_h { |k, v| [v, k] }.transform_keys(&:to_s)`, which the cop would allow, or just doing both in the single `to_h` block. Do you have a reason to prefer `map` here?

Reply:
The current implementation corrects the above example to invalid syntax:

```ruby
hash.transform_keys { |k| k.to_s } { |k, v| [v, k] }
```

However I think you're right that it's not a false positive, the autocorrect just needs to handle the second block correctly. I don't think we can safely reorder the operations in the general case though, since the second block may depend on the results of the first.